gr-qtgui/lib/matrix_sink_impl.cc

```diff
@@ -0,0 +1,146 @@
+                     gr::io_signature::make(0, 0, 0))
+{
+    d_vlen = vlen;
+    d_name = name;
+    d_parent = parent;
+    d_display = nullptr;
+    d_signal = nullptr;
+    d_argv = nullptr;
+    d_qApplication = nullptr;
+
+    if (qApp != nullptr) {
+        d_qApplication = qApp;
+    } else {
+        int argc = 1;
+        d_argv = new char;
+        d_argv[0] = '\0';
+        d_qApplication = new QApplication(argc, &d_argv);
+    }
+    d_display = new matrix_display(name,
+                                   num_cols,
+                                   vlen,
+                                   contour,
+                                   color_map,
+                                   interpolation,
+                                   0,
+                                   1,
+                                   0,
+                                   1,
+                                   0,
+                                   1,
+                                   "X",
+                                   "Y",
+                                   "Z",
+                                   d_parent);
+
+    d_signal = new matrix_display_signal();
+
+    qRegisterMetaType<QVector<double>>("QVector<double>");
+    QObject::connect(d_signal,
+                     &matrix_display_signal::data_ready,
+                     d_display,
+                     &matrix_display::set_data);
+}
+
+/*
+ * Our virtual destructor.
+ */
+matrix_sink_impl::~matrix_sink_impl()
+{
+
+    if (!d_display->isClosed()) {
+        d_display->close();
+    }
+}
+
+void matrix_sink_impl::exec_() { d_qApplication->exec(); }
+
+QWidget* matrix_sink_impl::qwidget() { return (QWidget*)d_display; }
+
+void matrix_sink_impl::set_x_axis_label(const std::string& x_axis_label)
+{
+    d_display->set_x_axis_label(x_axis_label);
+}
+
+void matrix_sink_impl::set_y_axis_label(const std::string& y_axis_label)
+{
+    d_display->set_y_axis_label(y_axis_label);
+}
+
+void matrix_sink_impl::set_z_axis_label(const std::string& z_axis_label)
+{
+    d_display->set_z_axis_label(z_axis_label);
+}
+
+void matrix_sink_impl::set_x_start(double x_start) { d_display->set_x_start(x_start); }
+
+void matrix_sink_impl::set_x_end(double x_end) { d_display->set_x_end(x_end); }
+
+void matrix_sink_impl::set_y_start(double y_start) { d_display->set_y_start(y_start); }
+
+void matrix_sink_impl::set_y_end(double y_end) { d_display->set_y_end(y_end); }
+
+void matrix_sink_impl::set_z_max(double z_max) { d_display->set_z_max(z_max); }
+
+void matrix_sink_impl::set_z_min(double z_min) { d_display->set_z_min(z_min); }
+
+
+int matrix_sink_impl::work(int noutput_items,
+                           gr_vector_const_void_star& input_items,
+                           gr_vector_void_star& output_items)
+{
+    auto in = static_cast<const input_type*>(input_items[0]);
+
+
+    QVector<double> qvec;
+    for (unsigned int i = 0; i < d_vlen; i++) {
+        qvec.append(in[i]);
+    }
```

Comment:
i tried that but since the incoming data is float so it wouldn't take that , also when casting the "in" from float to double the plot wasn't accurate so had to stick to that.